staging/src/k8s.io/apiserver/pkg/audit/policy/checker.go

```diff
@@ -0,0 +1,164 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package policy
+
+import (
+	"strings"
+
+	"k8s.io/apiserver/pkg/apis/audit"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+)
+
+const (
+	// DefaultAuditLevel is the default level to audit at, if no policy rules are matched.
+	DefaultAuditLevel = audit.LevelNone
+)
+
+// Checker exposes methods for checking the policy rules.
+type Checker interface {
+	// Check the audit level for a request with the given authorizer attributes.
+	Level(authorizer.Attributes) audit.Level
+}
+
+// NewChecker creates a new policy checker.
+func NewChecker(policy audit.Policy) Checker {
+	return &policyChecker{policy}
+}
+
+type policyChecker struct {
+	audit.Policy
+}
+
+func (p *policyChecker) Level(attrs authorizer.Attributes) audit.Level {
+	for _, rule := range p.Rules {
+		if ruleMatches(&rule, attrs) {
+			return rule.Level
+		}
+	}
+	return DefaultAuditLevel
+}
+
+// Check whether the rule matches the request attrs.
+func ruleMatches(r *audit.PolicyRule, attrs authorizer.Attributes) bool {
+	if len(r.Users) > 0 {
+		if !hasString(r.Users, attrs.GetUser().GetName()) {
+			return false
+		}
+	}
+	if len(r.UserGroups) > 0 {
+		matched := false
+		for _, group := range attrs.GetUser().GetGroups() {
+			if hasString(r.UserGroups, group) {
+				matched = true
+				break
+			}
+		}
+		if !matched {
+			return false
+		}
+	}
+	if len(r.Verbs) > 0 {
+		if !hasString(r.Verbs, attrs.GetVerb()) {
+			return false
+		}
+	}
+
+	if len(r.Namespaces) > 0 || len(r.Resources) > 0 {
```

Comment:
`ruleMatchesResource` checks whether the request matches any Namespaces or Resources from the rule. If neither are set, there's nothing to check :)